File: solution/regular-expression-matching.py
class Solution(object):

    def isMatch(self, s, p):
        return self.isMatchI(s, p, 0, 0)

    def isMatchI(self, s, p, si, pi):
        """
        :type s: str
        :type p: str
        :rtype: bool
        """
        if len(s) - si == 0:
            if len(p) - pi == 0 or len(p) - pi == 2 and p[pi + 1] == "*":
                return True
            else:
                return False
        if len(p) - pi == 0:
            return False

        if len(p) - pi >= 2 and p[pi + 1] == "*":
            idx = si
            while idx < len(s) and (s[idx] == p[pi + 0] or p[pi + 0] == "."):
                if self.isMatchI(s, p, idx, pi + 2):
                    return True
                idx += 1
                pass
            return self.isMatchI(s, p, si, pi + 2)
        else:
            if s[si] == p[pi] or p[pi] == ".":
                return self.isMatchI(s, p, si + 1, pi + 1)
            pass
        return False


s = Solution()
# print(s.isMatch("aa","a") == False)
# print(s.isMatch("aa","aa") == True)
# print(s.isMatch("aaa","aa") == False)
# print(s.isMatch("aa", "a*") == True)
# print(s.isMatch("aa", ".*") == True)
# print(s.isMatch("ab", ".*") == True)
# print(s.isMatch("aab", "c*a*b") == True)
print(s.isMatch("aaaaaaaaaaaaab", "a*a*a*a*a*a*a*a*a*a*c") == False)

File: solution/regular-expression-matching.cpp
#include <assert.h>
#include <iostream>
using namespace std;

class Solution {
public:
    bool isMatch(string s, string p) {
        return this->isMatch(s, p, 0, 0);
    }
    
    bool isMatch(string s, string p, int si, int pi) {
        if (pi >= p.length()) return si == s.length();

        // next char is not '*': must match current character
        if (pi + 1 >= p.length() || p[pi + 1] != '*') {
            return ((p[pi] == s[si] || p[pi] == '.') && this->isMatch(s, p, si+1, pi+1));
        }
        
        // next char is '*'
        while (si < s.length() && (p[pi] == s[si] || p[pi] == '.')) {
            if (this->isMatch(s, p, si, pi+2)) return true;
            si++;
        }
        return this->isMatch(s, p, si, pi+2);
    }
};

int main( void )
{
    Solution *s = new Solution();
    cout<<(s->isMatch("aa", "a*") == true)<<endl;
    cout<<(s->isMatch("aa","a") == false)<<endl;
    cout<<(s->isMatch("aa","aa") == true)<<endl;
    cout<<(s->isMatch("aaa","aa") == false)<<endl;
    cout<<(s->isMatch("aa", "a*") == true)<<endl;
    cout<<(s->isMatch("aa", ".*") == true)<<endl;
    cout<<(s->isMatch("ab", ".*") == true)<<endl;
    cout<<(s->isMatch("aab", "c*a*b") == true)<<endl;
    cout<<(s->isMatch("a", "ab*") == true)<<endl;
    
}